### PR TITLE
Write expectations from the proposal author explicitly

### PR DESCRIPTION
**Pull request by chshersh**
Dear CLC members,

I would like to offer some documentation improvements in the `PROPOSALS.md` document. The goal of these changes is the following:

+ **Lower the barrier** to creating a new proposal by documenting the process in more details
+ **Decrease the frustration** of proposal authors by writing explicitly what's expected from them
+ **Establish the expectations** from proposal authors around breaking changes

The specific changes are:

* 📚 A quick summary of the proposal in the beginning
* 🚧 Expectations from proposal authors
* 🏷️ Meanings of labels in the CLC repository

Let me know what you think!

This PR is more like an invitation to dialogue between CLC members and less like me telling others what to do. I've written my understanding of CLC processes, so please, share your opinion if you think differently 🙂 

[Rendered](https://github.com/chshersh/core-libraries-committee/blob/chshersh/clearer-proposal-expectations/PROPOSALS.md)

### CODE REVIEW

File: PROPOSALS.md
Comment (phadej) [inline PROPOSALS.md:71]: If 3. and 4. are indeed a hard requirement, then you'll need a label for proposals which ware blocked on that.

And maybe also a label / intermediate step of soft approval. I wouldn't go through the trouble of writing migration guide and especially compatibility patches if there it's unclear if the proposal will be accepted at all.

Comment (hasufell) [reply PROPOSALS.md:71]: I don't think these are hard requirements at all.

CLC has not established a baseline for these things yet, afaiu, so I'd say this document should be as general as possible without making assumptions about how CLC members will vote or what they will demand in order to vote +1.

Impact assessment is general enough that everyone can get behind, I think.

File: PROPOSALS.md
Comment (Bodigrim) [inline PROPOSALS.md:65]: The proposal process is described in more details at https://github.com/haskell/core-libraries-committee/blob/main/PROPOSALS.md. Should we merge these documents?..

Comment (chshersh) [reply PROPOSALS.md:65]: @Bodigrim Sorry, I didn't quite get your suggestion. Could you elaborate a bit more, please?

I already made my changes in the existing `PROPOSALS.md` document instead of creating a new one. Did you probably mean that it makes sense to merge `README.md` and `PROPOSALS.md`?

File: PROPOSALS.md
Comment (chshersh) [reply PROPOSALS.md:71]: @phadej That makes total sense 😌 I will change the wording to mention that you'll need to write a migration guide and submit patches only after your proposal is approved.

<hr>

@hasufell I agree with you and I don't think that having or not having impact assessment / migration guides / submitted patches should affect vote. However, after the impact assessment it may become clear that the proposed change breaks lots of stuff and this **actually may** affect the vote. So I would say that performing an impact assessment **is** hard requirement for breaking changes.

Comment (phadej) [reply PROPOSALS.md:71]: current proposals.md says differently though:

> Once everything that breaks receives a patch, let us know in a comment where they've been submitted. Then, the committee will be able to vote!

Implying patches need to be written prior the vote.

I guess, I'm not going to propose any breaking changes.

Comment (Bodigrim) [reply PROPOSALS.md:71]: Preparing _some sort of_ patches is a necessary part of impact assessment: we cannot assess the exact amount of breakage unless there is a patch which allows compilation to pass. However, such patches are allowed to be fairly crude and not suitable for merging (e. g., not backwards-compatible).

Migration guide indeed is required only after approval. 

There is a certain ambiguity around "submit" verb. Note the current wording is "let us know in a comment where they've been submitted", meaning that it's fine to submit patches as a gist somewhere, not necessarily raise PRs. This is intentional. Partly because sometimes a package has no public repository, or it's not on GitHub, or it's not even Git-based; partly because package maintainers cannot (or are usually unwilling to) assess a PR, providing compatibility with a future, unreleased version of `base`. It's important to retain this subtlety of meaning in the updated process documentation.

Comment (parsonsmatt) [reply PROPOSALS.md:71]: Yeah, I wouldn't want to have high barrier to writing proposals and trying to make improvements.

Certain changes with big negative impact would call for greater effort on the proposal author to ease that impact, but I wouldn't want someone to feel like they need to fork `base` and see how Stackage responds to a change *before* asking if it is a reasonable idea in the first place.

File: PROPOSALS.md
Comment (parsonsmatt) [inline PROPOSALS.md:12]: I think I'd rather have `Discuss and refine proposal` as an implied step between each of these - I would personally rather evaluate a small idea/proposal for feasibility before jumping into the next step, esp an impact assessment.

File: PROPOSALS.md
Comment (parsonsmatt) [inline PROPOSALS.md:66]: Let's note that these aren't required *up front* - possibly,

```suggestion
To be more precise, as the proposal author, you will be expected to do the following 
tasks, or find volunteers who'll do this for you. You may post a brief proposal and
request feedback on feasibility before investing time and effort in the full process.
```

**Review by parsonsmatt — COMMENTED** [review]
Overall looks like an improvement to have more information on process available, though I do want to clarify the amount of work required up front. Haskell is lazy, and we can be too 😄

**Comment by chshersh** [comment]
I believe, the suggested change by @parsonsmatt addresses the question of whether writing the migration guide or else is mandatory and when it's expected to do so.

If you think that wording in a specific paragraph can be improved, please let me know 🙂 

Otherwise, I believe this PR is ready, and I would like to merge it.

@mixphix @tomjaguarpaw @hasufell @angerman, please, share your feedback if you have any 🤗

**Comment by angerman** [comment]
I'm +1 on this. It improves the documentation, which is almost always good, as long as it's not misleading. My only concern is around the impact assessment. And this is rather tricky. We rightfully can't expect someone to do a proper full blown impact assessment without actually implementing it. On the other hand whether or not a proposal is promising is based on a rather accurate impact assessment. I fear we are missing some impact validation _after_ the implementation. What if the actual impact diverges signifiant from the theorised impact when the proposal was written? I doubt anyone would intentionally miss some impacts, but there seems to be no process for "Oh no! This change had a significantly larger impact than expected."

**Comment by Bodigrim** [comment]
> We rightfully can't expect someone to do a proper full blown impact assessment without actually implementing it. On the other hand whether or not a proposal is promising is based on a rather accurate impact assessment. I fear we are missing some impact validation after the implementation. What if the actual impact diverges signifiant from the theorised impact when the proposal was written?

On constrast to GHC Steering Committee, CLC proposals (except very trivial) must include a full implementation upfront, no need to theorize.

**Comment by hasufell** [comment]
I think this is fine, except the point about submitting patches to libraries. 

Whether that is a requirement or not has not been discussed properly and afaiu different members have different opinions on this, so it should be removed from the patch.

**Comment by angerman** [comment]
> [...] CLC proposals (except very trivial) must include a full implementation upfront [...]
@Bodigrim thank you for clarifying that; that however seems to disagree with this then:

```
 1. [Create an issue][open-issue] with your idea.
 2. Make [impact assessment](#impact-assessments) for breaking changes.
 3. Wait for the vote.
 4. If accepted, implement the proposal.
```

From your description it sounds more like:
```
 1. [Create an issue][open-issue] with your idea.
 2. Make [impact assessment](#impact-assessments) for breaking changes.
 3. Implement the proposal.
 4. Wait for the vote.
 5. Merge if accepted.
```
And then we can probably swap 2 and 3?

Maybe I'm completely misreading this though 😓

**Comment by chshersh** [comment]
> I think this is fine, except the point about submitting patches to libraries.

@hasufell I see that my wording didn't preserve the desired subtlety of showcasing breakage with patches, so I'll rewrite this part to be more explicit.

> I fear we are missing some impact validation after the implementation. What if the actual impact diverges signifiant from the theorised impact when the proposal was written?

@angerman I'm afraid, the Haskell community doesn't have the means to compile **all currently active** Haskell projects with the new changes to see the actual real impact. It's completely possible that Stackage shows no breakages while some private codebase will be completely obliterated by a small change if they rely on it too much.

I believe, the [Haskell Stability Group](https://github.com/haskellfoundation/stability) should figure out the policy for allowed breaking changes. But as CLC members, all we can do to minimise the impact of breakage is to be more cautious about accepting proposals with breaking changes and more upfront about soliciting early feedback from a wider community via various social media channels.

> And then we can probably swap 2 and 3?

It makes sense to me to see the implementation before voting on it. I'll change the wording in this list.

<hr>

From my understanding, if I read all the above comments right, the process should look like this:

```
1. Create an issue with your idea.
2. (When requested) Implement the proposal.
3. (If requested) Make impact assessment (it's mandatory for breaking changes).
4. (If requested) Let us know in a comment about patches to impacted libraries.
5. Wait for the vote.
6. (If accepted) Write a migration guide (for proposals with breaking changes).
```

**Comment by tomjaguarpaw** [comment]
> the [Haskell Stability Group](https://github.com/haskellfoundation/stability) should figure out the policy for allowed breaking changes

FYI the SWG is not a policy-making body. It is only a forum for interested parties to discuss issues related to stability with no authority whatsoever.

**Comment by chshersh** [comment]
I changed my documentation patch according to the latest remarks.

Now, the proposal process docs say:

* The implementation will be required at some point before the vote (even if the proposal won't be accepted)
* You're not expected to submit patches to all impacted libraries as Pull Requests but you may be required to tell about those patches and required migration changes, so CLC can see how difficult those patches are and it may impact the final vote

<hr>

Please, review again and let me know if anything else needs to be done before this can be merged 🙂

**Review by Bodigrim — APPROVED** [review]
LGTM!

**Review by mixphix — APPROVED** [review]

**Comment by Bodigrim** [comment]
Merging, this is certainly better than status quo.

**Pull request merged**
